framework/src/problems/FEProblemBase.C

```diff
@@ -6574,12 +6574,34 @@ FEProblemBase::checkNonlinearConvergence(std::string & msg,
   NonlinearSystemBase & system = getNonlinearSystemBase();
   MooseNonlinearConvergenceReason reason = MooseNonlinearConvergenceReason::ITERATING;
 
+  Real fnorm_old;
   // This is the first residual before any iterations have been done,
   // but after preset BCs (if any) have been imposed on the solution
   // vector.  We save it, and use it to detect convergence if
   // compute_initial_residual_before_preset_bcs=false.
   if (it == 0)
+  {
     system._initial_residual_after_preset_bcs = fnorm;
+    fnorm_old = fnorm;
+    _n_nl_pingpong = 0;
+  }
+  else
+    fnorm_old = system._last_nl_rnorm;
+
+  // ping pong will alwasy start with an increase in the residual value.
+  const bool increase = fnorm > fnorm_old ? true : false;
+
+  if (_n_nl_pingpong == 0 && increase)
+    _n_nl_pingpong += 1;
+  else if (_n_nl_pingpong == 0 && !increase)
+    _n_nl_pingpong = 0;
+  else
+  {
+    if ((_n_nl_pingpong % 2 == 1 && !increase) || (_n_nl_pingpong % 2 == 0 && increase))
+      _n_nl_pingpong += 1;
+    else
+      _n_nl_pingpong = 0;
```

Comment:
If I'm not crazy, I think these 4 lines give you all the logic you need

Reply:
yes, you are right.